File: src/main/clients/openrgb/effects/AbstractEffect.ts
import { AuraBrightness } from '@commons/models/Aura';
import { LoggerMain } from '@tser-framework/main';

import Device from '../client/classes/Device';
import { RGBColor } from '../client/classes/RGBColor';
import Client from '../client/client';

export abstract class AbstractEffect {
  protected isRunning = false;
  protected hasFinished = false;
  protected brightness = 0;
  protected color: RGBColor = RGBColor.fromHex('#FF0000');
  protected logger: LoggerMain;

  constructor() {
    this.logger = LoggerMain.for(this.getName());
  }

  public async start(
    client: Client,
    devices: Array<Device>,
    brightness: AuraBrightness,
    color: RGBColor
  ): Promise<void> {
    if (this.isRunning) {
      await this.stop();
    }

    this.hasFinished = false;
    if (brightness == AuraBrightness.OFF) {
      devices.forEach((dev, i) =>
        client.updateLeds(i, Array(dev.leds.length).fill(RGBColor.fromHex('#000000')))
      );
      this.hasFinished = true;
      return;
    } else {
      switch (brightness) {
        case AuraBrightness.LOW:
          this.brightness = 0.33;
          break;
        case AuraBrightness.MEDIUM:
          this.brightness = 0.67;
          break;
        case AuraBrightness.HIGH:
          this.brightness = 1;
          break;
      }

      this.color = color;
      this.logger.info('Starting effect');
      this.isRunning = true;

      ((): void => {
        this.applyEffect(client, devices);
      })();
    }
  }

  public async stop(): Promise<void> {
    if (this.isRunning) {
      this.logger.info('Stopping effect');
      this.isRunning = false;

      while (!this.hasFinished) {
        await new Promise<void>((resolve) => {
          setTimeout(() => {
            resolve();
          }, 100);
        });
      }
    }
  }

  protected abstract applyEffect(client: Client, devices: Array<Device>): void;

  public abstract getName(): string;
}

File: RogPerfTuner/src/clients/tcp/open_rgb/effects/abstract/abstract_effect.cpp
#include "../../../../../../include/clients/tcp/open_rgb/effects/abstract/abstract_effect.hpp"

#include <mutex>
#include <optional>
#include <string>
#include <thread>

#include "../../../../../../include/models/hardware/rgb_brightness.hpp"
#include "../../../../../../include/utils/enum_utils.hpp"
#include "../../../../../../include/utils/string_utils.hpp"
#include "../../../../../../include/utils/time_utils.hpp"

void AbstractEffect::start(const DeviceList& devices, const RgbBrightness& brightness) {
	if (_is_running) {
		stop();
	}

	if (brightness == RgbBrightness::OFF) {
		logger->info("Turning off RGB");
		for (auto& dev : devices) {
			_client.setDeviceColor(dev, Color::Black);
		}
		return;
	}

	if (!_color.has_value()) {
		logger->info("Starting effect '" + getName() + "' with " + StringUtils::toLowerCase(toName(brightness)) + " brightness");
	} else {
		logger->info("Starting effect '" + getName() + "' with " + StringUtils::toLowerCase(toName(brightness)) + " brightness and color " +
					 StringUtils::toUpperCase(_color.value().toHex()));
	}
	_brightness = toInt<RgbBrightness>(brightness) / 100.0;
	_is_running = true;

	_thread = std::thread([this, &devices] {
		this->_thread_main(devices);
	});
}

void AbstractEffect::stop() {
	if (_is_running) {
		logger->info("Stopping effect");
		_is_running = false;
		if (_thread.joinable()) {
			_thread.join();
		}
	}
}

void AbstractEffect::_sleep(const double& seconds) {
	double remaining = seconds * 1000;
	while (_is_running && remaining > 0) {
		double nap = std::min(100.0, remaining);
		TimeUtils::sleep(nap);
		remaining -= nap;
	}
}

void AbstractEffect::_set_colors(Device& dev, const std::vector<orgb::Color>& colors) {
	if (dev.enabled && _is_running) {
		std::lock_guard<std::mutex> lock(_mutex);
		if (_is_running) {
			std::vector<orgb::Color> dimmed_colors;
			for (const auto& c : colors) {
				dimmed_colors.push_back(c * _brightness);
			}
			_client.setDeviceColors(dev, dimmed_colors);
		}
	}
}

void AbstractEffect::_thread_main(const DeviceList& devices) {
	apply_effect(devices);
	logger->info("Effect finished");
}

AbstractEffect::AbstractEffect(Client& client, const std::string& name, const std::optional<std::string>& color)
	: Loggable(name), _name(name), _client(client) {
	auto parts = StringUtils::split(name, ' ');
	for (size_t i = 0; i < parts.size(); i++) {
		parts[i] = StringUtils::capitalize(parts[i]);
	}
	logger = LoggerProvider::getLogger(StringUtils::join(parts, "") + "Effect");
	if (color.has_value()) {
		_color = Color::fromRgb(*color);
	}
}

const std::string AbstractEffect::getName() {
	return _name;
}

bool AbstractEffect::supportsColor() {
	return _color.has_value();
}

void AbstractEffect::setColor(std::string color) {
	if (supportsColor()) {
		_color = Color::fromRgb(color);
	}
}

std::optional<std::string> AbstractEffect::getColor() {
	std::optional<std::string> res = std::nullopt;
	if (_color.has_value()) {
		res = _color->toHex();
	}
	return res;
}